feat(parser): add token start and end index/line/column to ParseError

3cp opened the pull request:
Meriyah actually reported the end position in ParseError, this change added tokenIndex/tokenLine/tokenColumn to ParseError as the start position, together with end position in index/line/column. User can catch the error from meriyah and then uses tool like @babel/code-frame to visualise the failing token.

closes #156

fisker commented:
Are you saying the error range should be "tokenIndex"to "index"?

I don't know, this seems confusing me. I didn't see other parser use those names either. Why not `.start`, `.end`?

3cp commented:
I agree, the naming needs cleanup.

I didn't want to introduce breaking change. It can be changed relatively easily.

Previously, we have `index`, `line`, `column` and `loc: { line, column }`. The `loc: {line, column }` is said to match acorn API.

For reference, there was a previous attempt #157 that changes to `loc: { start: { line, column }, end: { line, column } }`.

3cp commented:
I will have a look what other parsers report.

3cp commented:
Every parser reported in different way. Let's forget about acorn API shape.

I introduced a breaking change. Changed ParseError to expose

```
start: number
end: number;
range: [ number, number ]
loc: {
  start: { line: number, column: number },
  end: { line: number, column: number }
}
```

Just like what we have in AST node. This should be quite clear.

fisker commented:
How about introduce new loc just for errors

```
loc: {
  line: number, // Maybe non-enumerable getter?
  column: number,
  start: { line: number, column: number },
  end: { line: number, column: number }
}

3cp commented:
We don't have to match acorn api shape to have loc.line and loc.column. Breaking change I think it's acceptable.

Parsers do very differently to each other. For example, typescript only provides start and length without line/column.

Let's see what's natural for us.
